monai/transforms/utils.py

```diff
@@ -2051,5 +2058,46 @@ def has_status_keys(data: torch.Tensor, status_key: Any, default_message: str =
     return True, None
 
 
+def distance_transform_edt(
+    img: NdarrayOrTensor, sampling: None | float | list[float] = None, force_scipy: bool = False
+) -> NdarrayOrTensor:
+    """
+    Euclidean distance transform, either GPU based with CuPy / cuCIM
+    or CPU based with scipy.ndimage.
+    Choice depends on cuCIM being available or scipy can be forced with the ``force_scipy`` flag.
+
+    Note that the runtime running on the CPU may be really depending on the inputs size.
+
+    Args:
+        img: Input image on which the distance transform shall be run.
+            For CuPy shape must be ([channel, ]spatial_dim1, spatial_dim2 [, spatial_dim3]), so 2D, 3D and 4D input only.
+            If you need to run the transform on other shapes, use the ``force_scipy`` flag.
+            4D input gets passed channel-wise to cupy.
+        sampling: Spacing of elements along each dimension. If a sequence, must be of length equal to the input rank;
+            if a single number, this is used for all axes. If not specified, a grid spacing of unity is implied.
+        force_scipy: Force the CPU based scipy implementation of the euclidean distance transform
+    """
+    distance_transform_edt, has_cucim = optional_import(
+        "cucim.core.operations.morphology", name="distance_transform_edt"
+    )
+
+    if has_cp and has_cucim and not force_scipy:
+        img_ = convert_to_cupy(img)
+        if img_.ndim == 4:
```

Comment:
For a 4D Tensor, this will lead to different results depending on whether you use force_scipy is True or False. Or even more important depending on whether you have CUDA available or not. I think we either need to use this logic for both cases or not support any input with dims > 3 here at all (see https://github.com/Project-MONAI/MONAI/pull/6981#issuecomment-1725969539) .

Reply:
Hi @john-zielke-snkeos,

first of all thanks for the comments :)

Yeah that is definitely a concern.
However by default the choice by default (without the force_scipy) only dependent on the import and will default to cupy which will simply crash if the input does not have the correct format.
We could at least make it more clear by enhancing the comment

`Note that the calculations of CuPy and Scipy can deviate, for details look into the cuCIM about distance_transform_edt(). Also for 4D input the CuPy version will calculate the distance transform channel-wise which returns different results compared to scipy.  If you need a full 4D distance transform, use scipy instead. We recommend to stick with one of the two libraries for the entire program.`

Do you think that would be enough? I don't see many other solutions right now apart from adding an actual working 4D implementation.

Reply:
Also to be on the save side, I did a trial with a 4D tensor, the results actually are different.

```python
(monai_dev) matteo@matteo-ubuntu-server:~/code/MONAI_matteo$ python
Python 3.10.11 (main, May 16 2023, 00:28:57) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from monai.transforms.utils import distance_transform_edt
>>> import numpy as np
>>> t = np.array([[[[1,1,1], [1,1,0], [0,0,0]]], [[[1,1,1], [1,1,0], [1,1,0]]],])
>>> t.ndim
4
>>> distance_transform_edt(t)
array([[[[2.       , 1.4142135, 1.       ],
         [1.       , 1.       , 0.       ],
         [0.       , 0.       , 0.       ]]],


       [[[2.236068 , 1.4142135, 1.       ],
         [2.       , 1.       , 0.       ],
         [2.       , 1.       , 0.       ]]]], dtype=float32)
>>> distance_transform_edt(t, force_scipy=True)
array([[[[2.        , 1.41421356, 1.        ],
         [1.        , 1.        , 0.        ],
         [0.        , 0.        , 0.        ]]],


       [[[2.23606798, 1.41421356, 1.        ],
         [1.41421356, 1.        , 0.        ],
         [1.        , 1.        , 0.        ]]]])

```